helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+
+This implementation will necessarily have some limitations and restrictions that the client Helm does not; namely, it will not be able to handle large charts because there is a 1M limit to CRD size.
+
+It is also assumed that the controller model may not choose to implement a number of the client features, such as dry runs, repository management, chart creation, etc.
+
+## Appendix B: What Is a Package Manager?
+
+If Helm is "the package manager for Kubernetes", we need to have a common explanation of what we mean by "package manager".
+
+Helm's notion of package management is informed by _operating system package managers_ such as:
+ - [APT](https://www.debian.org/doc/manuals/debian-reference/ch02.en.html)
+ - [RPM](http://rpm.org/documentation.html)
+ - [Homebrew](https://docs.brew.sh/)
+ - [APK](https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management)
+ - [Portage](https://wiki.gentoo.org/wiki/Portage)
+ - [Snap](https://docs.snapcraft.io/core/)
+
+It is important to not overzealously compare Helm to programming language library package managers, as such systems have a fundamentally different goal.
+
+Package managers such as Apt, Homebrew, and RPM provide the following services:
+
+- They define a package format
+- They define a package repository architecture
+- They define a security model (signing, verification, etc)
+- They specify a programming langugage (or languages) to be used for management logic (e.g. shell scripts, Ruby)
+- They manage the installation, upgrade, and deletion of a package
+- They provide dependency resolution features and version management
+- They expose configuration options that the user may override
+- They start services (e.g. when installing a database, the database is started) and often provide tooling for administrators to manage those services (e.g. systemd/upstart/sysV init scripts)
+- They provide tools to search, list, and discover packages
+- They implement (or make it easy to implement) environment-specific (os, processor, distro) optimizations
+- They provide an opinion as to how to best construct and maintain packages
+- They surface documentation (typically via output during installation)
+
+Package managers tend to involve all of these because it is important in package management to keep external dependencies to a minimum.
+
+Consequently, if Helm is to fit into such a paradigm, it is natural that Helm will provide at least most, if not all of these.
+
+Helm is not a configuration management tool (such as Chef or Puppet). Helm does provide certain features of such systems, though. Helm's templating and configuration model is a common feature of configuration managers (and also of package managers).
+
+## Appendix C: Alternatives to Embedding Lua Engine
+
+There were two main choices to be made about how to extend Helm with a better lifecycle event model:
+
+1. Do we load lifecycle hooks as extensions, as part of a chart, or both?
+2. What extension mechanism do we support?
```

Comment:
Option 3, plugin running in cluster.

k8s can handle multiple arch/os options today.  It also will behave identically on each users machine while lua may behave quite differently if admin 1 and admin 2 run it based on various machine differences like mac/windows/linux, what their home directory structure looks like, etc.